lightning/src/ln/channelmanager.rs

```diff
@@ -5077,28 +5181,81 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
 	fn forward_htlcs(&self, per_source_pending_forwards: &mut [(u64, OutPoint, Vec<(PendingHTLCInfo, u64)>)]) {
 		for &mut (prev_short_channel_id, prev_funding_outpoint, ref mut pending_forwards) in per_source_pending_forwards {
 			let mut forward_event = None;
+			let mut new_intercept_events = Vec::new();
+			let mut failed_intercept_forwards = Vec::new();
 			if !pending_forwards.is_empty() {
 				let mut forward_htlcs = self.forward_htlcs.lock().unwrap();
-				if forward_htlcs.is_empty() {
-					forward_event = Some(Duration::from_millis(MIN_HTLC_RELAY_HOLDING_CELL_MILLIS))
-				}
+				let short_to_chan_info = self.short_to_chan_info.read().unwrap();
 				for (forward_info, prev_htlc_id) in pending_forwards.drain(..) {
-					match forward_htlcs.entry(match forward_info.routing {
-							PendingHTLCRouting::Forward { short_channel_id, .. } => short_channel_id,
-							PendingHTLCRouting::Receive { .. } => 0,
-							PendingHTLCRouting::ReceiveKeysend { .. } => 0,
-					}) {
+					let scid = match forward_info.routing {
+						PendingHTLCRouting::Forward { short_channel_id, .. } => short_channel_id,
+						PendingHTLCRouting::Receive { .. } => 0,
+						PendingHTLCRouting::ReceiveKeysend { .. } => 0,
+					};
+					let forward_htlcs_empty = forward_htlcs.is_empty();
+					match forward_htlcs.entry(scid) {
 						hash_map::Entry::Occupied(mut entry) => {
-							entry.get_mut().push(HTLCForwardInfo::AddHTLC { prev_short_channel_id, prev_funding_outpoint,
-							                                                prev_htlc_id, forward_info });
+							entry.get_mut().push(HTLCForwardInfo::AddHTLC(PendingAddHTLCInfo {
+								prev_short_channel_id, prev_funding_outpoint, prev_htlc_id, forward_info }));
 						},
 						hash_map::Entry::Vacant(entry) => {
-							entry.insert(vec!(HTLCForwardInfo::AddHTLC { prev_short_channel_id, prev_funding_outpoint,
-							                                             prev_htlc_id, forward_info }));
+							if !short_to_chan_info.contains_key(&scid) &&
+							   forward_info.amt_incoming.is_some() &&
+							   fake_scid::is_valid_intercept(&self.fake_scid_rand_bytes, scid)
+							{
+								let intercept_id = InterceptId(Sha256::hash(&forward_info.incoming_shared_secret).into_inner());
+								let mut pending_intercepts = self.pending_intercepted_payments.lock().unwrap();
+								match pending_intercepts.entry(intercept_id) {
+									hash_map::Entry::Vacant(entry) => {
+										new_intercept_events.push(events::Event::PaymentIntercepted {
```

Comment:
@ViktorTigerstrom pointed out [here](https://github.com/lightningdevkit/rust-lightning/pull/1601#discussion_r925036528) that an attacker could force us to break backwards compat by brute forcing a valid intercept scid and attempting to forward over it. Should we make this whole thing opt-in?

Reply:
I personally think it would make sense to make this opt-in for now :).

Is general forward interception planned to be opt-in as well? If that's the case then I definitely think this should be opt-in, and replaced with the config flag for general forward interception once implemented.
If we plan to make general forward interception the default, then it's probably unnecessary to make this opt-in.

Reply:
Made this opt-in due to the great catch of breaking compatibility. General interception will also be opt-in via a separate knob, as it would make sense for some users to only care about one type of interception or another.

Reply:
Ok makes sense :)